docs: add Azure initialization steps

dsloanm opened the pull request:
Adds initialization steps for the Azure cloud. Includes:

* setting up credentials
* bootstrapping a controller
* clean-up

This is for the machine cloud - nothing on Azure Kubernetes (AKS) yet.

dsloanm commented:
Latest commit adds Azure Kubernetes (AKS) steps. This is needed for COS and the identity stack but you can get away with just the machine cloud when deploying only the Slurm charms.

dsloanm commented:
Newest commit moves the cleanup instructions to their own how-to rather than having them in the middle of initialization steps.

dsloanm commented:
Last commit adds a note for registering the "Microsoft.Compute" resource provider. If this isn't registered, the VM quotas page gives you the message: "The selected provider is not registered for some of the selected subscriptions".

AshleyCliff commented on howto/cleanup-cloud-resources.md:
```diff
@@ -0,0 +1,105 @@
+az group delete -n <resourceGroup>
+:::
+
+Should destroying the controller or AKS instance take a long time or be seemingly stuck, proceed to delete resources manually [via the Azure web portal](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/manage-resources-portal) or [via the `az` CLI](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/manage-resources-cli). To list any remaining Azure  ...
```
Is there a rough estimate of how long a 'long time' would be? Is it ok for them to wait or do they _have_ to go finish the process manually?

dsloanm replied:
I've added a suggestion that the user wait 15 minutes before moving onto the manual process. Test deployments usually clean up within 10 minutes but can take longer. It's typically OK to wait but rarely the manual process is necessary when things are clearly stuck, e.g. the destroy command has been repeating its cleaning up message for over an hour.

AshleyCliff reviewed (changes requested):
Excellent work getting this all written out! I've added some thoughts and comments to tighten some areas up.

dsloanm commented on howto/initialize-cloud-environment.md:
```diff
@@ -138,6 +138,130 @@ Machine  State    Address        Inst id        Base          AZ  Message
+If the Azure Portal page for adjusting VM quota appears blank or contains the message "The selected provider is not registered for some of the selected subscriptions", confirm that [the *Microsoft.Compute* resource provider is registered](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-providers-and-types) for your subscription.
+:::
+
+:::{hint}
+To decide on suitable VMs, it may be useful to refer to [Sizes for virtual machines in Azure](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/overview). A typical Charmed HPC deployment will use a mix of high-performance and GPU-accelerated compute VMs for cluster compute nodes, and general purpose VMs for other node types.
+:::
```
Agreed - moved this to a normal paragraph and added it to the compatibility matrix PR.

dsloanm commented on howto/initialize-cloud-environment.md:
```diff
@@ -138,6 +138,130 @@ Machine  State    Address        Inst id        Base          AZ  Message
+To decide on suitable VMs, it may be useful to refer to [Sizes for virtual machines in Azure](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes/overview). A typical Charmed HPC deployment will use a mix of high-performance and GPU-accelerated compute VMs for cluster compute nodes, and general purpose VMs for other node types.
+:::
+
+### Add Azure cloud to Juju
```
Yep, agreed that's better. Done!

dsloanm commented on howto/initialize-cloud-environment.md:
```diff
@@ -138,6 +138,130 @@ Machine  State    Address        Inst id        Base          AZ  Message
+
+You will be asked to authenticate the requests via your web browser with the following message:
+
+:::{code-block} shell
+To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code <auth code> to authenticate.
+:::
```
Good call. I've added a line break for clarity.

dsloanm commented on howto/initialize-cloud-environment.md:
```diff
@@ -138,6 +138,130 @@ Machine  State    Address        Inst id        Base          AZ  Message
+
+In a web browser, open the [authentication page](https://microsoft.com/devicelogin), sign in as required, and enter `<auth code>` from the terminal output.
+
+You will be asked to authenticate twice, to allow creation of two different resources in Azure.
```
Done.

dsloanm commented on howto/initialize-cloud-environment.md:
```diff
@@ -138,6 +138,130 @@ Machine  State    Address        Inst id        Base          AZ  Message
+
+Once the credentials have been added successfully, the following message will be displayed:
+
+:::{code-block} shell
+Credential <your credential name> added locally for cloud "azure".
+:::
```
This one I think makes sense as `<your credential name>` since the message will contain the custom credential name entered by the user and it keeps it consistent with other uses further down the doc.

Alternatively, we could replace all instances with `<credential-name>` to avoid having both it and `<your credential name>` (and similarly replace `<Azure subscription ID>` with `subscription-id`, etc.). Let me know what you think.

AshleyCliff replied:
Might be best to replace all with `<credential-name>` - the 'your' part feels unnecessary, and can these credentials have spaces in the names? Especially for names that _should_ be a single word or at least contain no spaces, using dashes or underlines in the generic variable helps preserve that pattern. I believe that's the pattern we follow for naming the models as well. Thoughts?

dsloanm replied:
On reflection, I've replaced most of these instances with concrete examples, e.g. `my-az-credential` rather than `<credential-name>`, since I think it's a bit clearer.

Definitely agreed that dashes are the way to go when using generic variables, however. I've switched the `az role assignment` sample command to use `<application-object-id>` and `<subscription-id>`.

Let me know your thoughts.

AshleyCliff replied:
This looks good to me - the 'example' naming and dashes both.

dsloanm commented on howto/initialize-cloud-environment.md:
```diff
@@ -138,6 +138,130 @@ Machine  State    Address        Inst id        Base          AZ  Message
+
+### Clean up
+
+:::{note}
+Always clean Azure resources that are no longer necessary! Abandoned resources are tricky to detect and can become expensive over time.
+:::
```
Done.

dsloanm commented on howto/initialize-cloud-environment.md:
```diff
@@ -191,6 +315,67 @@ charmed-hpc-k8s  1        default  k8s
+az aks get-credentials --resource-group aks --name ${JUJU_NAME} --context aks
+:::
+
+This will add your AKS credentials to the file `~/.kube/config`. Now, add the AKS cloud to the controller:
```
I've removed `${JUJU_NAME}` and `--context aks` here, rather than have the user make another naming decision. The cluster context will just default to `charmed-aks-cluster` in the `~/.kube/config` file without a `--context` specified.

dsloanm commented on howto/cleanup-cloud-resources.md:
```diff
@@ -0,0 +1,105 @@
+the Kubernetes (K8s) cloud, hosting control plane applications.
+
+:::{note}
+Always clean up public cloud resources that are no longer necessary! Abandoned resources are tricky to detect and can become expensive over time.
+:::
+
+:::{warning}
+Clean-up may result in **permanent data loss**. Ensure all data you wish to preserve has been migrated to a safe location before proceeding.
+:::
```
I've merged these boxes and made the data loss warning the first paragraph for emphasis.

dsloanm commented on howto/cleanup-cloud-resources.md:
```diff
@@ -0,0 +1,105 @@
+az aks list
+:::
+
+Look for the values of `"name"` and `"resourceGroup"` then delete both the instance and its resource group with:
```
I've added edited example output as a guide. I've avoided including the full output since it's quite a few lines of uninterpreted JSON...

dsloanm commented on howto/cleanup-cloud-resources.md:
```diff
@@ -0,0 +1,105 @@
+(howto-cleanup-cloud-resources)=
+# How to clean up cloud resources
+
+This guide demonstrates how to clean up {ref}`a previously initialized cloud environment <howto-initialize-cloud-environment>`
+and a deployed Charmed HPC cluster. This includes clean-up of both the machine cloud, hosting compute plane applications, and
+the Kubernetes (K8s) cloud, hosting control plane applications.
```
Reworded this - let me know what you think.

AshleyCliff commented on howto/initialize-cloud-environment.md:
```diff
@@ -138,6 +138,130 @@ Machine  State    Address        Inst id        Base          AZ  Message
+
+### Bootstrap Azure cloud controller
+
+First, set the default region for deploying Azure instances, including the controller itself. Refer to [What are Azure regions?](https://learn.microsoft.com/en-us/azure/reliability/regions-overview) for an overview of available regions. To set the default region to East US:
```
Considering how integrated it is within the rest of the section, the description should probably stay there and just have a similar version in the other places it'd be relevant. This might be a good use for the 'reuse/' pieces @NucciTheBoss has been trying out.

dsloanm replied:
On a second pass, I think what I had was too "Explanation-y" for a how-to. I've cut it down to just 'the Juju cloud environment can now be initialized or "bootstrapped"' and kept the link to the `juju bootstrap` command documentation. We'll probably want similar in other cloud sections but maybe not scope for a 'reuse/' piece at the moment. Thoughts?

AshleyCliff replied:
Good point on the Explanation vs How-to. Maybe a small description of 'bootstrapping' goes in the Explanation section and gets linked here - I'll add to the docs issues.

AshleyCliff commented:
Great updates! A few areas to review where we've got discussions going

dsloanm commented on howto/initialize-cloud-environment.md:
```diff
@@ -191,6 +315,67 @@ charmed-hpc-k8s  1        default  k8s
+:::{code-block} shell
+export JUJU_NAME=aks-$USER-$RANDOM
+az aks create -g aks -n ${JUJU_NAME} --enable-managed-identity --node-count 1 --node-vm-size=Standard_D4s_v4 --generate-ssh-keys
+:::
```
For this, I've gone with a note that "VM sizes should be chosen to accommodate control plane applications like identity management and observability services" and deferred to the Azure documentation for deploying and sizing AKS.

AshleyCliff replied:
I agree that's likely the best way to approach this - refer to the Azure docs when possible.

dsloanm commented:
I've completed a second pass and added responses to the discussions. I've also rebased on the current `main`.

The CI is currently failing as links to the Juju docs are 404ing. Will take a look at updating them.
